### DISCUSSION TITLE
Notes on maintainer checkins

### DISCUSSION THREAD

**Discussion by connortann** [shap/shap#3737] [General]
This is a place for rough notes on our informal maintainer checkins. In future we may move this to a Wiki.

**Comment by connortann** [comment]
## 2024-06-28

Connor & @CloseChoice 

Tobias's priorities:
 
 - fixing LSTMs for DeepTF/Kernelexplainer
 - increasing test coverage (also adding tests for js)
 - improving or adding docstrings (there are a lot of functions without docstrings and it is not really obvious what they do)
 
Potential upcoming features:
 
 - Writing a Deepexplainer/Kernelexplainer for Jax
 - Standardising visualisation API

Other:

- Keeping up regular release cadence
- Ongoing effort to welcome new contributors and maintainers
- Adopting type checking

**Comment by connortann** [comment]
## 2024-07-11 (catch up with Scott)

@connortann , @slundberg & @thatlittleboy 

EDIT: added summary notes from the discussion

### Review of past year, current state of affairs

- Focussing heavily on bug fixes and improving stability
- Trying to improve reliability of the test & build: regression tests, more automation
- Keen to grow the pool of contributors and maintainers

### Scott's recap of shap's dev history

- Very early on: started with model agnostic approach for explanations. Just KernelExplainer
- After that came TreeExplainer. Primarily compute optimisation (but exponential!)
- Integration with LightGBM and XGBoost. To some extent catboost, though they did that by themselves
- Some work from NVidea: @RAMitchell contributed the GPUTreeExplainer code. 
- There was a group that did Text and Image explaining work. Not integrated in the same way as TreeExplainer, so not as high performance.

### Thoughts on possible future features

- Lots of active research in interpretability. LLMs obviously a main area of focus at present.
- Recent colleague of Scott investigating accelerating SHAP for language models. Similar to PartitionExplainer approach
- Perhaps do the same for vLLM and llama.cpp, as is already done for HuggingFace
  - Hesitancy is that the model agnostic and tree explainer are clear value-adds, took many months to get the core algorithms working well. It's no small feat to do this work.
  - Make it community driven - perhaps give some clear direction. "It would be cool if" - e.g. partition explainer for text were tightly integrated with these LLM API endpoints.
- Could potentially happen in one of two ways:
  1. It could simply just call these LLM endpoints.
  2. More as a research thing, this could be made a lot quicker, more optimised. Anything that makes it faster and more optimised is valuable.

### New vs legacy API

Envisaged plan to fully migrate to Explanation interface:

- Should be fine. Just don't want to break people's code. Err towards long deprecation cycles. As long as it doesn't make the code too ugly.
- Background on the change:
  - Initially it was good to return something that you don’t' have to ask question about
  - Then the plots need all these arguments
  - And particularly hard to slice stuff. Slicer was made together with InteretML, working with GAMs. Scott's impression is that the interpretML didn't have the momentum to maintain slicer, so it's now primarily used by shap.

### shap.benchmark

- Connor: module is not linted or tested, but kept around for now. It would be good to fix it up in future.
- Nice to have, but probably not the top priority at the moment
- Context on the original purpose:
  - There has been a lot of work around feature attribution explanation methods
  - There are loads of papers around on why certain methods are better
  - Benchmark was made to show how this actually plays out over a wide group of metrics.
  - For example, if you consider the feature "if you turn it off, it will make the biggest difference" - then you can just turn them off in order. 
  - So, what you really need is a broad benchmark of metrics. There's a trade-off that you can always hyper-optimiser for one task.
- Leave as a "nice-to-have" for future
- Jeremy: Perhaps it should be spun off as a separate repo, within the `shap` org.

### Testing & reliability

- Did Scott have a way to test the GPU code on CI?
  - Might be a path with Nvidea to create a CI pipeline?
  - Hesitation is that there might be some friction to integrate with existing GH Actions process
- Handling packaging on conda-forge
 - Currently not very smooth, tests often fail at release time

### Miscellany

- Was there any significance behind the occasional use of `issubclass(type(foo), Bar)` rather than `isinstance(foo, Bar`)?
  - Not really, probably safe to replace
- Rationale behind the OpChain stuff: fun with custom metaclasses
  - The key idea is to have a neat way to represent the intent to perform a set of dot-chained operations. The metaclass approach works; perhaps it could be simplified to just regular class methods.
- Plans for the PartitionExplainer? There is a lot of commented out code and functions that are never used, e.g. [owen3](https://github.com/shap/shap/blob/master/shap/explainers/_partition.py#L395). I guess this is supposed to be a performance upgrade to the standard owen function, but some more information would be great.
  - Did not cover

**Reply by CloseChoice** [reply]
Nice, looks good. Just one miscellany point I have (and added) for now.

**Reply by connortann** [reply]
Pinging @dsgibbons & @thatlittleboy in case either of you are around and might like to join?

**Reply by thatlittleboy** [reply]
Sure, would love to if the timing is right, thanks for the ping @connortann Have you settled on a time yet with Scott?

**Reply by connortann** [reply]
Great to hear from you @thatlittleboy ! We're meeting 9:00 PST / 16:00 GMT, which is the earliest Scott can make. I'll forward on an invite; I appreciate the time zone difference is problematic. In any case, feel free to edit the list above or comment with any topics, and I can write up notes afterwards.

**Reply by thatlittleboy** [reply]
Cool, I'll try to make it, I'm particularly interested in hearing about the OpChain and the more legacy / less utilized parts of the package.
